### PR TITLE
Feature/implement user microservice

### PR DESCRIPTION
**Pull request by LoneXDii**

### CODE REVIEW

File: UserService/UserService.Domain/Exceptions/ApiException.cs
Comment (MefistoSamael) [inline UserService/UserService.Domain/Exceptions/ApiException.cs:1]: move all exceptions to application

File: UserService/UserService.API/Middleware/ExceptionsMiddleware.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+			StatusCode = ex switch
+			{
+				LoginException => StatusCodes.Status400BadRequest,
+				InvalidTokenException => StatusCodes.Status404NotFound,
+				BadRequestException => StatusCodes.Status400BadRequest,
+				NotFoundException => StatusCodes.Status404NotFound,
+				RegisterException => StatusCodes.Status400BadRequest,
+				_ => StatusCodes.Status500InternalServerError
+			}
+		};
```
Comment (MefistoSamael) [inline UserService/UserService.API/Middleware/ExceptionsMiddleware.cs]: why don't you use ProblemDetails?

File: UserService/UserService.Application/UseCases/AuthUseCases/EmailConfirmationUseCase/EmailConfirmationRequestHandler.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+{
+	public async Task Handle(EmailConfirmationRequest request, CancellationToken cancellationToken)
+	{
+		var user = await userManager.FindByEmailAsync(request.email);
+		if (user is null)
+		{
+			throw new NotFoundException("No user with such email");
+		}
+
+		var code = HttpUtility.UrlDecode(request.code);
```
Comment (MefistoSamael) [inline UserService/UserService.Application/UseCases/AuthUseCases/EmailConfirmationUseCase/EmailConfirmationRequestHandler.cs]: Application layer have to know nothing about fact that you are using web. It is better to remove this from application layer

File: UserService/UserService.Application/UseCases/AuthUseCases/LoginUserUseCase/LoginUserRequestHandler.cs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+internal class LoginUserRequestHandler(SignInManager<AppUser> signInManager, UserManager<AppUser> userManager, ITokenService tokenService)
+	: IRequestHandler<LoginUserRequest, TokensDTO>
+{
+	public async Task<TokensDTO> Handle(LoginUserRequest request, CancellationToken cancellationToken)
+	{
+		var signInResult = await signInManager.PasswordSignInAsync(request.LoginModel.Email, request.LoginModel.Password, false, false);
+
+		if (!signInResult.Succeeded)
+		{
+			throw new LoginException("Incorrect email or password");
```
Comment (MefistoSamael) [inline UserService/UserService.Application/UseCases/AuthUseCases/LoginUserUseCase/LoginUserRequestHandler.cs]: use badRequest instead

File: UserService/UserService.Application/UseCases/AuthUseCases/LoginUserUseCase/LoginUserRequestHandler.cs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+		}
+
+		var user = await userManager.FindByEmailAsync(request.LoginModel.Email);
+
+		if (!user!.EmailConfirmed)
+		{
+			throw new LoginException("Email is not verified");
```
Comment (MefistoSamael) [inline UserService/UserService.Application/UseCases/AuthUseCases/LoginUserUseCase/LoginUserRequestHandler.cs]: use badRequest

File: UserService/UserService.Application/UserService.Application.csproj
Comment (MefistoSamael) [inline UserService/UserService.Application/UserService.Application.csproj:1]: dependency rule violated. Application layer have to not know about infrastructure

File: UserService/UserService.Application/UseCases/AuthUseCases/RegisterUserUseCase/RegisterUserRequestHandler.cs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+	{
+		var user = mapper.Map<AppUser>(request.RegisterModel);
+
+		if(request.RegisterModel.Avatar is not null)
+		{
+			using Stream stream = request.RegisterModel.Avatar.OpenReadStream();
+			var imageId = await blobService.UploadAsync(stream, request.RegisterModel.Avatar.ContentType);
+
+			//Later will be replace by Ocelot endpoint
+			var imageUrl = $"http://127.0.0.1:10000/devstoreaccount1/avatars/{imageId}";
```
Comment (MefistoSamael) [inline UserService/UserService.Application/UseCases/AuthUseCases/RegisterUserUseCase/RegisterUserRequestHandler.cs]: don't configure url here. Do it in blobService

File: UserService/UserService.Application/UseCases/AuthUseCases/RegisterUserUseCase/RegisterUserRequestHandler.cs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+			var code = await userManager.GenerateEmailConfirmationTokenAsync(user);
+			code = HttpUtility.UrlEncode(code);
+			var email = HttpUtility.UrlEncode(user.Email);
+			var confirmationLink = $"https://localhost:7001/api/account/confirm/email={email}&code={code}";
+			await emailService.SendEmailConfirmationCodeAsync(user.Email, confirmationLink);
+		}
+		else
+		{
+			var errors = JsonSerializer.Serialize(result.Errors);
+			throw new RegisterException($"Cannot register user: {errors}");
```
Comment (MefistoSamael) [inline UserService/UserService.Application/UseCases/AuthUseCases/RegisterUserUseCase/RegisterUserRequestHandler.cs]: use badRequest

File: UserService/UserService.Application/UseCases/AuthUseCases/RegisterUserUseCase/RegisterUserRequestHandler.cs [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+			user.AvatarUrl = imageUrl;
+		}
+
+		var result = await userManager.CreateAsync(user, request.RegisterModel.Password);
+		if (result.Succeeded)
+		{
+			var code = await userManager.GenerateEmailConfirmationTokenAsync(user);
+			code = HttpUtility.UrlEncode(code);
+			var email = HttpUtility.UrlEncode(user.Email);
+			var confirmationLink = $"https://localhost:7001/api/account/confirm/email={email}&code={code}";
```
Comment (MefistoSamael) [inline UserService/UserService.Application/UseCases/AuthUseCases/RegisterUserUseCase/RegisterUserRequestHandler.cs]: this shouldn't be located in application

File: UserService/UserService.Application/UseCases/AuthUseCases/RegisterUserUseCase/RegisterUserRequestHandler.cs [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+			await emailService.SendEmailConfirmationCodeAsync(user.Email, confirmationLink);
```
Comment (MefistoSamael) [inline UserService/UserService.Application/UseCases/AuthUseCases/RegisterUserUseCase/RegisterUserRequestHandler.cs]: all await statements should be followed and preceeded by blank lines

File: UserService/UserService.Application/UseCases/UserUseCases/GetUserInfoUseCase/GetUserInfoRequestHandler.cs [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+	public async Task<UserInfoDTO> Handle(GetUserInfoRequest request, CancellationToken cancellationToken)
+	{
+		var user = await userManager.FindByIdAsync(request.userId);
+        if (user is null)
+        {
+            throw new NotFoundException("No such user");
+        }
+
+        var userDTO = mapper.Map<UserInfoDTO>(user);
+        return userDTO;
```
Comment (MefistoSamael) [inline UserService/UserService.Application/UseCases/UserUseCases/GetUserInfoUseCase/GetUserInfoRequestHandler.cs]: return statement should be preceeded by blank line

File: UserService/UserService.Application/UseCases/UserUseCases/GetUserInfoUseCase/GetUserInfoRequestHandler.cs [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+internal class GetUserInfoRequestHandler(UserManager<AppUser> userManager, IMapper mapper)
+	: IRequestHandler<GetUserInfoRequest, UserInfoDTO>
+{
+	public async Task<UserInfoDTO> Handle(GetUserInfoRequest request, CancellationToken cancellationToken)
+	{
+		var user = await userManager.FindByIdAsync(request.userId);
+        if (user is null)
+        {
+            throw new NotFoundException("No such user");
+        }
```
Comment (MefistoSamael) [inline UserService/UserService.Application/UseCases/UserUseCases/GetUserInfoUseCase/GetUserInfoRequestHandler.cs]: conditional statements should be followed and preceded by blank line. Fix tabulation

File: UserService/UserService.Application/UseCases/UserUseCases/ListUsersWithPaginationUseCase/ListUsersWithPaginationRequestHandler.cs [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+	: IRequestHandler<ListUsersWithPaginationRequest, PaginatedListModel<UserInfoDTO>>
+{
+	public async Task<PaginatedListModel<UserInfoDTO>> Handle(ListUsersWithPaginationRequest request, CancellationToken cancellationToken)
+	{
+		int count = await userManager.Users.CountAsync();
+		var totalPages = (int)Math.Ceiling(count / (double)request.pageSize);
+
+		if(request.pageNo > totalPages)
+		{
+			throw new NotFoundException("No such page");
```
Comment (MefistoSamael) [inline UserService/UserService.Application/UseCases/UserUseCases/ListUsersWithPaginationUseCase/ListUsersWithPaginationRequestHandler.cs]: fix tabulation

File: UserService/UserService.Application/UseCases/UserUseCases/ListUsersWithPaginationUseCase/ListUsersWithPaginationRequestHandler.cs [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace UserService.Application.UseCases.UserUseCases.ListUsersWithPaginationUseCase;
+
+internal class ListUsersWithPaginationRequestHandler(UserManager<AppUser> userManager, IMapper mapper)
+	: IRequestHandler<ListUsersWithPaginationRequest, PaginatedListModel<UserInfoDTO>>
+{
+	public async Task<PaginatedListModel<UserInfoDTO>> Handle(ListUsersWithPaginationRequest request, CancellationToken cancellationToken)
+	{
+		int count = await userManager.Users.CountAsync();
+		var totalPages = (int)Math.Ceiling(count / (double)request.pageSize);
```
Comment (MefistoSamael) [inline UserService/UserService.Application/UseCases/UserUseCases/ListUsersWithPaginationUseCase/ListUsersWithPaginationRequestHandler.cs]: use validators to check input values

File: UserService/UserService.Application/UseCases/UserUseCases/UpdateEmailUseCase/UpdateEmailRequestHandler.cs [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+
+internal class UpdateEmailRequestHandler(UserManager<AppUser> userManager, IEmailService emailService)
+	: IRequestHandler<UpdateEmailRequest>
+{
+	public async Task Handle(UpdateEmailRequest request, CancellationToken cancellationToken)
+	{
+		if (!Regex.IsMatch(request.newEmail, @"^\S+@\S+\.\S+$"))
+		{
+			throw new BadRequestException("Incorrect email");
+		}
```
Comment (MefistoSamael) [inline UserService/UserService.Application/UseCases/UserUseCases/UpdateEmailUseCase/UpdateEmailRequestHandler.cs]: move this to validator

File: UserService/UserService.Application/UseCases/UserUseCases/UpdateEmailUseCase/UpdateEmailRequestHandler.cs [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+
+		user.Email = request.newEmail;
+		user.UserName = request.newEmail;
+		user.EmailConfirmed = false;
+		await userManager.UpdateAsync(user);
+
+		var code = await userManager.GenerateEmailConfirmationTokenAsync(user);
+		code = HttpUtility.UrlEncode(code);
+		var email = HttpUtility.UrlEncode(user.Email);
+		var confirmationLink = $"https://localhost:7001/api/account/confirm/email={email}&code={code}";
```
Comment (MefistoSamael) [inline UserService/UserService.Application/UseCases/UserUseCases/UpdateEmailUseCase/UpdateEmailRequestHandler.cs]: this shouldn't be located in application layer

File: UserService/UserService.Application/UseCases/UserUseCases/UpdateUserUseCase/UpdateUserRequestHandler.cs [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+			{
+				var avatarId = user.AvatarUrl.Split("/").Last();
+				await blobService.DeleteAsync(new Guid(avatarId));
+			}
+
+			using Stream stream = request.userDTO.Avatar.OpenReadStream();
+			var imageId = await blobService.UploadAsync(stream, request.userDTO.Avatar.ContentType);
+
+			//Later will be replace by Ocelot endpoint
+			var imageUrl = $"http://127.0.0.1:10000/devstoreaccount1/avatars/{imageId}";
```
Comment (MefistoSamael) [inline UserService/UserService.Application/UseCases/UserUseCases/UpdateUserUseCase/UpdateUserRequestHandler.cs]: this shouldn't be located in application

File: UserService/UserService.Application/Mapping/AppMappingProfile.cs [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+internal class AppMappingProfile : Profile
+{
+	public AppMappingProfile()
+	{
+		CreateMap<RegisterDTO, AppUser>()
+			.ForMember(dest => dest.UserName, opt => opt.MapFrom(src => src.Email));
+
+		CreateMap<AppUser, UserInfoDTO>();
+
+		CreateMap<UpdateUserDTO, AppUser>();
```
Comment (MefistoSamael) [inline UserService/UserService.Application/Mapping/AppMappingProfile.cs]: separate mappings into different files

File: UserService/UserService.Infrastructure/Database/DbInitializer/DbInitializer.cs [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+		AppUser customerUser = new()
+		{
+			UserName = "client1@gmail.com",
+			Email = "client1@gmail.com",
+			EmailConfirmed = true,
+			FirstName = "Customer",
+			LastName = "Customer"
+		};
+		await _userManager.CreateAsync(customerUser, "Qwe123*");
+	}
```
Comment (MefistoSamael) [inline UserService/UserService.Infrastructure/Database/DbInitializer/DbInitializer.cs]: use hasData methods

File: UserService/UserService.Infrastructure/Services/Authentication/ITokenService.cs [hunk 18]
```diff
@@ -0,0 +1,12 @@
...
+using UserService.Infrastructure.Models;
+
+namespace UserService.Infrastructure.Services.Authentication;
+
+public interface ITokenService
+{
+    Task<TokensDTO> GetTokensAsync(AppUser user);
+    Task<string> RefreshAccessTokenAsync(string refreshToken);
+    Task InvalidateRefreshTokenAsync(AppUser user);
+	Task<string> GetRefreshTokenAsync(AppUser user);
```
Comment (MefistoSamael) [inline UserService/UserService.Infrastructure/Services/Authentication/ITokenService.cs]: fix tabulation

File: UserService/UserService.Infrastructure/Services/Authentication/TokenService.cs [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+	{
+		user.RefreshTokenValue = Guid.NewGuid().ToString();
+		user.RefreshTokenExpiresAt = DateTime.Now.AddDays(Convert.ToDouble(_configuration["JWT:RefreshExpireDays"]));
+		await _userManager.UpdateAsync(user);
+
+		return user.RefreshTokenValue;
+	}
+
+	private async Task<string> GetAccessTokenAsync(AppUser user)
+    {
```
Comment (MefistoSamael) [inline UserService/UserService.Infrastructure/Services/Authentication/TokenService.cs]: fix tabulation

**Review by MefistoSamael — CHANGES_REQUESTED** [review]

File: UserService/UserService.Application/UserService.Application.csproj
Comment (LoneXDii) [reply UserService/UserService.Application/UserService.Application.csproj:1]: Discussed

**Review by MefistoSamael — APPROVED** [review]

**Pull request merged**